log error when restoring cache from file

rshayer95 opened the pull request:
### What this PR does

log error returned by `fastcache.LoadFromFile` before falling back to creating a new cache instance. this improves observability and helps detect problems like file corruption or permission issues early.

this replaces `fastcache.LoadFromFileOrNew` with a custom function `loadFromFileOrNewWithLog` that explicitly logs errors encountered during cache restoration.

---

### Related Issue

Closes #8934

---

### Test Plan

- manually tested by simulating a missing file scenario  
- ensured expected log output on cache load failure  
- verified normal cache creation fallback path  

---

### Changelog

log error when cache fails to restore from file during workingsetcache initialization (#8934)

---

### Checklist

- [x] Signed commits  
- [x] Follows coding and commit message conventions  
- [x] Tested manually  
- [x] Scope limited to relevant change  
- [x] Changelog entry added

makasim commented on lib/workingsetcache/cache.go:
```diff
@@ -64,8 +66,29 @@ func Load(filePath string, maxBytes int) *Cache {
+//
+// Return:
+// - *fastcache.Cache: A loaded or newly created cache instance.
+func loadFromFileOrNewWithLog(filepath string, maxBytes int) *fastcache.Cache {
```
loadFromFileOrNew should be sufficient.

Could you add a changelog entry ?

Could you add some tests as well?
I'm not sure we have to log every case. For instance, on the first start, the file might not exist, resulting in a "file not found" error. We shouldn't log that as an error.

hagen1778 replied:
It seems like the only one actionable error in cache load is:
```
			return nil, fmt.Errorf("cache file %s contains maxBytes=%d; want %d", filePath, maxBytes, expectedBucketChunks*chunkSize*bucketsCount)
```

It happens when available memory has changed (limits changed or flags were overriden). In this case, vmstorage will discrad the prev cache and start from scratch.

This error could be actionable only in case if user by mistake changed memory allowance or some other misconfig is affecting this. 

Other cases or other errors returned from this function seems non-actionable to me. Hence, no need in pritinting them.

---------

Another case I could think of is pointing vmstorage cache path to temporary directory that doesn't persist between pod restarts. In this case, it is easier to log just `creating new cache` line instead of `can't open` errors. Maybe, `creating new cache`  is sufficient enough for identifying the previously described case with memory allowance.

rshayer95 commented:
I have made the necessary modifications. Kindly review the changes and provide feedback to ensure they meet the specified requirements.

makasim commented on lib/workingsetcache/cache.go:
```diff
@@ -78,17 +80,25 @@ func Load(filePath string, maxBytes int) *Cache {
+	} else if !errors.Is(err, os.ErrNotExist) && !strings.Contains(err.Error(), "no such file or directory") {
+		// Only log if it's not a missing file
+		logger.Infof("creating new cache at %s due to: %v", filePath, err)
+	}
```
I think we should log the other cases as errors, as they likely indicate format corruption or regression issues.

rshayer95 replied:
this one is done

rshayer95 commented on lib/workingsetcache/cache.go:
```diff
@@ -78,17 +80,25 @@ func Load(filePath string, maxBytes int) *Cache {
+	// Inverted logic: handle unexpected/loggable cases first
+	if strings.Contains(err.Error(), "contains maxBytes") {
+		logger.Warnf("cache file %s has mismatched memory size: %v", filePath, err)
+	} else if !errors.Is(err, os.ErrNotExist) && !strings.Contains(err.Error(), "no such file or directory") {
```
made necessary changes and they are working fine as expected with new requirements

rshayer95 commented:
Implemented enhanced error logging in loadFromFileOrNew to log all unexpected errors during cache file loading, improving visibility into cache load failures. Adjusted error handling to distinguish between missing files and other errors. Updated related tests to ensure coverage of these cases.

Please review the changes and let me know if any further improvements or fixes are needed.

makasim commented:
@hagen1778 @zekker6 @rtm0 please review

hagen1778 commented on lib/workingsetcache/cache.go:
```diff
@@ -64,8 +68,32 @@ func Load(filePath string, maxBytes int) *Cache {
+// and a new cache is created with the specified maxBytes size.
+//
+// This improves observability by surfacing cache loading issues that would otherwise be silently ignored.
+func loadFromFileOrNew(filePath string, maxBytes int) *fastcache.Cache {
+	cache, err := fastcache.LoadFromFile(filePath, maxBytes)
+	if errors.Is(err, os.ErrNotExist) {
+		logger.Infof("cache file %s not found; init new cache", filePath)
+		return fastcache.New(maxBytes)
+
+		// covers the cache reset due to max memory size change at
+		// https://github.com/VictoriaMetrics/fastcache/blob/198c85ee90a1f65127126b5904c191e70f083cbf/file.go#L133
+	} else if err != nil && strings.Contains(err.Error(), "contains maxBytes") {
+		logger.Warnf("cache file %s couldn't be used: %s; init new cache", filePath, err)
+		return fastcache.New(maxBytes)
+	} else if err != nil {
+		logger.Errorf("cache file %s invalid: %s; init new cache", filePath, err)
+		return fastcache.New(maxBytes)
+	}
+
+	return cache
+}
```
```suggestion
func loadFromFileOrNew(filePath string, maxBytes int) *fastcache.Cache {
	cache, err := fastcache.LoadFromFile(filePath, maxBytes)
	if err == nil {
		return cache
	}
	
	switch {
	case errors.Is(err, os.ErrNotExist):
		logger.Infof("cache file %s not found", filePath)
	case strings.Contains(err.Error(), "contains maxBytes"):
		logger.Warnf("cache file %s couldn't be used: %s", filePath, err)
	default:
		logger.Errorf("cache file %s invalid: %s", filePath, err)
	}
	
	logger.Infof("initing new cache file %s", filePath)
	return fastcache.New(maxBytes)
}
```

makasim replied:
I'm not a big fan of doing this:

```golang
if err == nil {
    // success path
}
```

It doesn’t feel idiomatic in Go (as I understand it). It’s easy to confuse == with !=, and since err != nil is used in 99% of cases, people are more likely to mentally read it as such.

I prefer when the success path flows straight downward, and all edge cases or errors branch off to the right. Yes, my approach introduces some duplication — like returning values multiple times or adding extra err != nil checks — and that’s not ideal, but I’m okay with that trade-off.

Also attaching the blog post on this topic: https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

func25 replied:
Partially agree with @makasim. It's not idiomatic but having `err == nil` like this is somewhat accepted as a team convention in this project.

makasim replied:
I've chosen to use the approach proposed by @hagen1778 for the following reasons:

- The majority seem to favour this approach, including @valyala, considers this approach to be the most effective.
- It's widely adopted in Go source code — according to a GitHub search, there are approximately [700 matches](https://github.com/search?q=repo%3Agolang%2Fgo%20%22err%20%3D%3D%20nil%22&type=code) for `err == nil` and around [2,000](https://github.com/search?q=repo%3Agolang%2Fgo+%22err+%21%3D+nil%22&type=code) for `err != nil`.

hagen1778 commented on lib/workingsetcache/cache_test.go:
```diff
@@ -0,0 +1,137 @@
+)
+
+func TestLoadFromFileOrNew(t *testing.T) {
+	initValidCache := func(t *testing.T, maxBytes int) string {
```
```suggestion
	newCache := func(t *testing.T, maxBytes int) string {
```

rtm0 commented on lib/workingsetcache/cache_test.go:
```diff
@@ -0,0 +1,137 @@
+
+func TestLoadFromFileOrNew(t *testing.T) {
+	newCache := func(t *testing.T, maxBytes int) string {
+		tmpDir, err := os.MkdirTemp("", "test")
+		if err != nil {
+			t.Fatal(err)
+		}
+		filePath := filepath.Join(tmpDir, "TestLoadFromFileOrNew.workingsetcache")
```
Then this could be rewrtitten as:

```go
path := filepath.Join(t.Name(), "workingsetcache")
```

makasim replied:
fixed

makasim commented on lib/workingsetcache/cache.go:
```diff
@@ -64,8 +68,32 @@ func Load(filePath string, maxBytes int) *Cache {
+		logger.Infof("cache file %s not found; init new cache", filePath)
+		return fastcache.New(maxBytes)
+
+		// covers the cache reset due to max memory size change at
```
fixed

makasim commented on lib/workingsetcache/cache.go:
```diff
@@ -64,8 +68,32 @@ func Load(filePath string, maxBytes int) *Cache {
+// This improves observability by surfacing cache loading issues that would otherwise be silently ignored.
+func loadFromFileOrNew(filePath string, maxBytes int) *fastcache.Cache {
+	cache, err := fastcache.LoadFromFile(filePath, maxBytes)
+	if errors.Is(err, os.ErrNotExist) {
```
I've chosen Roman's approach, see https://github.com/VictoriaMetrics/VictoriaMetrics/pull/8952#discussion_r2114037761

makasim commented on lib/workingsetcache/cache_test.go:
```diff
@@ -0,0 +1,137 @@
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
+)
+
+func TestLoadFromFileOrNew(t *testing.T) {
```
fixed

makasim commented on lib/workingsetcache/cache_test.go:
```diff
@@ -0,0 +1,137 @@
+		return filePath
+	}
+
+	t.Run("CacheDirNotExist", func(t *testing.T) {
```
test added

makasim commented on lib/workingsetcache/cache_test.go:
```diff
@@ -0,0 +1,137 @@
+			t.Fatal("expected a new cache instance, got nil")
+		}
+
+		if len(cache.Get(nil, []byte("foo"))) != 0 {
```
fixed

rtm0 commented on lib/workingsetcache/cache_test.go:
```diff
@@ -4,51 +4,69 @@ import (
 
+	newCache := func(t *testing.T, maxBytes int) string {
 		c := New(maxBytes)
+		t.Cleanup(c.Stop)
```
You need to stop it after the creation, otherwise you will have 2 instances of cache using the same filepath with their own goroutines potentially messing with those files. t.Cleanup() will stop it only once the test case finishes.

rtm0 reviewed (approved):
LGTM, thanks!

hagen1778 commented on lib/workingsetcache/cache.go:
```diff
@@ -64,8 +68,31 @@ func Load(filePath string, maxBytes int) *Cache {
+// If loading fails due to an error (e.g. corrupted or unreadable file), the error is logged
+// and a new cache is created with the specified maxBytes size.
+//
+// This improves observability by surfacing cache loading issues that would otherwise be silently ignored.
```
```suggestion
```
In my opinion, justifying error handling in function description seems out of place.
I think it is better to add comment to the error handling itself.

hagen1778 commented on lib/workingsetcache/cache.go:
```diff
@@ -64,8 +68,31 @@ func Load(filePath string, maxBytes int) *Cache {
+	}
+
+	if errors.Is(err, os.ErrNotExist) {
+		logger.Infof("cache at path %s missing files; init new cache", filePath)
```
all cases below contain `init new cache` message. Would it be simpler to log this message unconditionally with `info` level?

hagen1778 commented on lib/workingsetcache/cache.go:
```diff
@@ -64,8 +68,31 @@ func Load(filePath string, maxBytes int) *Cache {
+		// https://github.com/VictoriaMetrics/fastcache/blob/198c85ee90a1f65127126b5904c191e70f083cbf/file.go#L133
+		logger.Warnf("%s; init new cache", err)
+	} else {
+		logger.Errorf("cache at path %s unusable: %s; init new cache", filePath, err)
```
```suggestion
		logger.Errorf("cache at path %s is invalid: %s; init new cache", filePath, err)
```

hagen1778 commented on lib/workingsetcache/cache_test.go:
```diff
@@ -0,0 +1,160 @@
+	"github.com/VictoriaMetrics/fastcache"
+)
+
+func TestLoadFromFileOrNew(t *testing.T) {
```
wdyt about re-writing tests in f-like approach? 
https://itnext.io/f-tests-as-a-replacement-for-table-driven-tests-in-go-8814a8b19e9e

It would align with most of other tests in the project.
I.e. see f-like approach below. Note, it doesn't test OK case, as it should be tested in a separate test function: `TestLoadFromFileOrNewOk` and `TestLoadFromFileOrError`

```
f := func(path string, expErr string) {
		logBuffer := &bytes.Buffer{}
		logger.SetOutputForTests(logBuffer)
		defer logger.ResetOutputForTest()

		cache := loadFromFileOrNew(path, 10000)
		if cache == nil {
			t.Fatal("expected a new cache instance, got nil")
		}
		if !strings.Contains(logBuffer.String(), "missing files; init new cache") {
			t.Fatalf("expected log message not found; got: %s", logBuffer.String())
		}
	}
	f("emptyPath", "missing files", 0)

	path := filepath.Join(t.Name(), "workingsetcache")
	if err := os.MkdirAll(path, 0777); err != nil {
		t.Fatalf("failed to create cache directory: %v", err)
	}
	f(path, "missing files", 0)

	path = newCache(t, 10000)
	if err := os.Remove(filepath.Join(path, `metadata.bin`)); err != nil {
		t.Fatalf("failed to remove metadata.bin file: %v", err)
	}
	f(path, "missing files", 0)

	path = newCache(t, 10000)
	if err := os.WriteFile(filepath.Join(path, `metadata.bin`), []byte(""), 0644); err != nil {
		t.Fatalf("failed to write test metadata file: %v", err)
	}
	f(path, "cannot read maxBucketChunks", 0)

	cachePath := newCache(t, 87654321)
	f(path, "contains maxBytes=10000", 0)
```

hagen1778 commented:
Thank you!
